### PR TITLE
Add option to hide plane widget vector

### LINKED ISSUE #908: Disable / hide normal vectors of the clip plane
Hi everyone,

I am wondering that it is possible to disable or hide the normal vectors of the clip plane ?

I need this method because, when I apply set_scale to the plotter, the corresponding normal vector is scaled as well and it will be distorted.

In addition, sometime we do not want users to rotate the clip plane, only slide them along the axis.

### PR DESCRIPTION
**Pull request by banesullivan**
Resolve #908

### CODE REVIEW

**Comment by adeak** [comment]
The changes look good to me. I have a few suggestions but I can't add them in review because they are too far from your changes... (ugh)

  1. I believe https://github.com/pyvista/pyvista/blob/2b4bda85fd0851363b7cd94186a3920247b9f6cd/pyvista/plotting/widgets.py#L324 is being fixed now
  2. The docstring should be amended [here](https://github.com/pyvista/pyvista/blob/2b4bda85fd0851363b7cd94186a3920247b9f6cd/pyvista/plotting/widgets.py#L384) and [here](https://github.com/pyvista/pyvista/blob/2b4bda85fd0851363b7cd94186a3920247b9f6cd/pyvista/plotting/widgets.py#L456)
  3. Adding a test case for the new kwarg would probably account for the missing coverage.
  4. The new feature might be handy enough to add among https://docs.pyvista.org/examples/03-widgets/plane-widget.html

**Comment by banesullivan** [comment]
Thanks for the feedback, @adeak! I will go back and resolve those things accordingly

**Review by Ilyas888 — APPROVED** [review]

**Comment by banesullivan** [comment]
This should be ready to go now.

Note that I added a feature where the normal vector is forcibly disabled anytime `assign_to_axis` is set

File: pyvista/plotting/widgets.py [hunk 1]
```diff
@@ -365,6 +381,40 @@ def add_mesh_clip_plane(self, mesh, normal='x', invert=False,
...
+            placed at the given bounds. Only valid when using an implicit
+            plane.
+
+        origin_translation : bool
+            If ``False``, the plane widget cannot be translated by its origin
+            and is strictly placed at the given origin. Only valid when using
+            an implicit plane.
+
+        implicit : bool
+            When ``True``, a ``vtkImplicitPlaneWidget`` is ued and when
```
Comment (adeak) [inline pyvista/plotting/widgets.py:410]:
```suggestion
            When ``True``, a ``vtkImplicitPlaneWidget`` is used and when
```

File: pyvista/plotting/widgets.py [hunk 2]
```diff
@@ -435,6 +487,37 @@ def add_mesh_slice(self, mesh, normal='x', generate_triangles=False,
...
```
Comment (adeak) [inline pyvista/plotting/widgets.py:512]:
```suggestion
            When ``True``, a ``vtkImplicitPlaneWidget`` is used and when
```

File: tests/test_widgets.py [hunk 3]
```diff
@@ -54,6 +54,18 @@ def test_widget_plane():
     p.add_plane_widget(callback=func, pass_widget=True, implicit=False)
     p.close()
 
+    p = pyvista.Plotter()
+    func = lambda normal, origin: normal # Does nothing
+    p.add_mesh(mesh)
+    p.add_plane_widget(callback=func, assign_to_axis='z', implicit=True)
+    p.close()
```
Comment (adeak) [inline tests/test_widgets.py:61]: Do we need the `NO_PLOTTING` skips when there's no `p.show()`, only `p.close()`? I never know :)

**Review by adeak — CHANGES_REQUESTED** [review]
There are two minor typos, but otherwise LGTM :)

**Comment by adeak** [comment]
Sorry, I don't know why half the comment was posted twice...

File: tests/test_widgets.py (hunk 3 above)
Comment (banesullivan) [reply tests/test_widgets.py:61]: Yes, we still need it regardless. If the system doesn't support plotting (no virtual display or graphics lib) then instantiating the plotter will fail.

A render window is created regardless of the `show()` call (show simply displays what already exists)

**Review by adeak — APPROVED** [review]

File: tests/test_widgets.py (hunk 3 above)
Comment (adeak) [reply tests/test_widgets.py:61]: I see, thanks for the explanation :)

**Comment by akaszynski** [comment]
I'm seeing a weird CI failure on `doctest`:

```
_____________ [doctest] pyvista.core.filters.PolyDataFilters.clean _____________
3805         --------
3806         Create a mesh with a degenerate face and then clean it,
3807         removing the degenerate face
3808 
3809         >>> import pyvista as pv
3810         >>> points = np.array([[0, 0, 0], [0, 1, 0], [1, 0, 0]])
3811         >>> faces = np.array([3, 0, 1, 2, 3, 0, 3, 3])
3812         >>> mesh = pv.PolyData(points, faces)
3813         >>> mout = mesh.clean()
3814         >>> print(mout.faces)
Expected:
    [3 0 1 2]
Got:
    [3 0 1 2 3 0 3 4]
```

I'll rerun the test as this might be just a flaky test.  Reporting here so we can recall which doctest example is flaky so we can consider adding a skip to the doctest check.

**Comment by akaszynski** [comment]
Looks like rerunning fixed it.  If this happens again we can skip testing the docstring as this seems to be an internal VTK issue.

**Review by akaszynski — APPROVED** [review]
Nice work.  Thanks @banesullivan

**Comment by adeak** [comment]
Two remarks, looking at the intermittently breaking doctest:

Firstly, I'm surprised it runs without `import numpy as np` :) I guess doctests have memory from earlier tests in the same class or module.

Secondly, that test is indeed too flaky to keep as is, I think. Running in ipython:
```py
>>> from itertools import count
... import numpy as np
... import pyvista as pv
... points = np.array([[0, 0, 0], [0, 1, 0], [1, 0, 0]])
... faces = np.array([3, 0, 1, 2, 3, 0, 3, 3])
...
... for i in count():
...     mesh = pv.PolyData(points, faces)
...     mout = mesh.clean()
...     if not np.array_equal(mout.faces, [3, 0, 1, 2]):
...         print(f'Surprising result {mout.faces} in iteration {i}...')

Surprising result [3 0 1 2 3 0 3 4] in iteration 8...
Surprising result [3 0 1 2 3 0 3 4] in iteration 25...
Surprising result [3 0 1 2 3 0 3 4] in iteration 41...
Surprising result [3 0 1 2 3 0 3 4] in iteration 53...
Surprising result [3 0 1 2 3 0 3 4] in iteration 55...
Surprising result [3 0 1 2 3 0 3 4] in iteration 63...
Surprising result [3 0 1 2 3 0 3 4] in iteration 92...
Surprising result [3 0 1 2 3 0 3 4] in iteration 121...
Surprising result [3 0 1 2 3 0 3 4] in iteration 124...
[...]
```
But based on a few tries this usually only happens when this block is run a second (or further) time...

**Comment by adeak** [comment]
Is there anything needed to be done here? It seems good to go to me.

**Comment by akaszynski** [comment]
> Is there anything needed to be done here? It seems good to go to me.

Now that you have approvals you're good to merge.

**Comment by adeak** [comment]
Technically it's @banesullivan who has the approvals, but you don't have to tell me twice :smile:

**Comment by leminhson** [comment]
@banesullivan : Thank you so much. The option to hide the plane widget vector is usefull.
However, if I create a plane widget which will slide along the x-axis and set the scale for plotter such as: **p.set_scale(1, 1, 50)** then it is really hard to pick up the plane for sliding along the x-axis.
Maybe this problem is due to the distortion of the plane widget vector (although it is hidden) along the z-axis?

**Comment by adeak** [comment]
@leminhson is this actually related to the hidden vector? Isn't the same issue there when the vector is shown? If so, then this is an unrelated issue, and you should probably raise that separately. I guess even if it's related to this new feature it might be better to open a new issue, explaining your problem with a small example.

**Comment by leminhson** [comment]
@adeak : Thank you for your suggestion.
I found that the real issue is not by show/hiding the normal vector but by a very large value of a scale factor. I will open a new thread for this issue.

**Pull request merged**